proxy/m2_actions.c

```diff
@@ -1186,6 +1186,101 @@ action_m2_container_propdel (struct req_args_s *args, struct json_object *jargs)
 }
 
 static enum http_rc_e
+_m2_container_snapshot(struct req_args_s *args, struct json_object *jargs)
+{
+	GError *err = NULL;
+	gchar **urlv = NULL;
+	gchar **urlv_snapshot = NULL;
+	char type[64] = {};
+	_get_meta2_realtype(args, type, sizeof(type));
+	const char *target_account, *target_container;
+	char target_cid[65] = {};
+	target_account = oio_url_get(args->url, OIOURL_ACCOUNT);
+	target_container = oio_url_get(args->url, OIOURL_USER);
+	strncpy(target_cid, oio_url_get(args->url, OIOURL_HEXID),
+			sizeof(target_cid)-1);
+	err = hc_resolve_reference_service(resolver, args->url, NAME_SRVTYPE_META2,
+			&urlv);
+	if (!err && (!urlv || !*urlv)) {
+		err = NEWERROR(CODE_CONTAINER_NOTFOUND, "No service located");
+	}
+	if (err)
+		goto cleanup;
+
+	struct json_object *jaccount = NULL;
+	struct json_object *jcontainer = NULL;
+	struct oio_ext_json_mapping_s m[] = {
+		{"account", &jaccount, json_type_string, 1},
+		{"container", &jcontainer, json_type_string, 1},
+		{NULL, NULL, 0, 0}
+	};
+
+	err = oio_ext_extract_json(jargs, m);
+	if (err)
+		goto cleanup;
+
+	const gchar *account = json_object_get_string(jaccount);
+	const gchar *container = json_object_get_string(jcontainer);
+	if(!strcmp(account, target_account) && !strcmp(container, target_container)) {
+		err = BADREQ("the snapshot should have a different account or reference");
+		goto cleanup;
+	}
+
+	oio_url_set(args->url, OIOURL_ACCOUNT, account);
+	oio_url_set(args->url, OIOURL_USER, container);
+	guint8 id[32] = {};
+	gchar hexid[65] = {};
+	oio_str_hash_name(id, NULL, account, container);
+	oio_str_bin2hex(id, sizeof(id),hexid, sizeof(hexid));
+	oio_url_set(args->url, OIOURL_HEXID, hexid);
```

Comment:
You should not need the previous 5 lines. Instead, just call `oio_url_set(args->url, OIOURL_HEXID, NULL)` that will reset the hexid currently stored. Then the subsequent call to fetch it will recompute it.